de.metas.contracts/src/main/java/de/metas/contracts/invoicecandidate/FlatrateTermInvoiceCandidateHandler.java

```diff
@@ -162,10 +175,11 @@ private I_C_Invoice_Candidate createCandidateForTerm(final I_C_Flatrate_Term ter
 
 		ic.setQtyOrdered(qty);
 
+		setOrderAndOrderLineIfNeeded(term, ic);
```

Comment:
imho the code would be much more cleaner and readable without this method.
It's used only in one place, you could inline it.
First question which pops in my mind when i read this is: set the order and order line IDs from where to where. who is who?

IMHO this is an example of taking the clean code book advises to the other extreme.... 
@metas-ts  wdyt?

Reply:
The method de.metas.contracts.invoicecandidate.FlatrateTermInvoiceCandidateHandler.createCandidateForTerm(I_C_Flatrate_Term) is quite big already, so why not split it? IMHO, we could split more in there, but I didn't want to be carried away.
And why inline it? For me, at least, is more clear explicit, but I guess is a matter of taste.
Maybe a better name, instead of setOrderAndOrderLineIfNeeded would be inheritOrderAndOrderLineIfNeededToInvoiceCandidate.

Reply:
maybe yes... and also usually the "to" is first argument
... or this is how we got used in C world :)

Reply:
ihmo: 
> First question which pops in my mind when i read this is: set the order and order line IDs from where to where. who is who?

I think we need to make a difference depending on what "this" is:
* seeing the method in the method outline: yes, agree (but note that it's a private method which therefore is only used in the "inner" context of that class)
* on the other hand, reading the method `createCandidateForTerm()` (maybe having started reading that class with the public `createCandidatesFor(..)` method), when it see `setOrderAndOrderLineIfNeeded(term, ic)` I think i could guess that since the whole method is about creating an invoice candidate from a term, the ic's order and order line are set from the terms ("if needed")...now if i wanted, I could dive into that method, or stay on the current level.

> IMHO this is an example of taking the clean code book advises to the other extreme....

afiu it's not taking the clean code book to the extreme, but just following the book's advise **partially**.
Afaiu the book advises to keep the methods short and not to mix different "levels" within one method..by extracting the new code into a different level, cris implements the issue requirement with
* minimizing the "short" rule's further violation (the method is aready too long) 
* and introducing a new "mixed-level" vialotion (by not breaking up the other details ontop methods as well)

> It's used only in one place, you could inline it.

imho the DRY principle is not the only reason to have dedicated methods..clarity alone can also be one.

from here i see two topics for discussion
* do you agree with what i wrote about the book? should I look up the passages so we can discuss more to the point on whether it's really meant that way and on whether we agree with them?
* should we in future avoid following the book "partially", at least on this low level?

Reply:
> Afaiu the book advises to keep the methods short and not to mix different "levels" within one method..

i think u nail it.